klippy/extras/z_thermal_adjust.py

```diff
@@ -0,0 +1,195 @@
+
+# Adjusts Z position in real-time using a thermal probe to e.g. compensate
+# for thermal expansion of the printer frame.
+
+import threading
+
+class ZThermalAdjuster:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+        self.lock = threading.Lock()
+        self.config = config
+
+        # Get config parameters, convert to SI units where necessary
+        self.temp_coeff = config.getfloat('temp_coeff', minval=-1, maxval=1,
+            default=0)
+        self.temp_sensor_name = config.get('temp_sensor')
+        self.smooth_time = config.getfloat('smooth_time', 2., above=0.)
+        self.inv_smooth_time = 1. / self.smooth_time
+
+        self.off_above_z = config.getfloat('z_adjust_off_above', 99999999.)
+        self.max_z_adjust_mm = config.getfloat('max_z_adjustment', 99999999.)
+
+        # Register printer events
+        self.printer.register_event_handler("klippy:connect",
+                                            self.handle_connect)
+        self.printer.register_event_handler("homing:home_rails_end",
+                                            self.handle_homing_move_end)
+
+        # Setup temperature sensor
+        self.last_temp = self.smoothed_temp = 0.
+        self.last_temp_time = 0.
+        self.ref_temperature = 0.
+
+        # Z transformation
+        self.z_adjust_mm = 0.
+        self.last_z_adjust_mm = 0.
+        self.adjust_enable = True
+        self.last_position = [0., 0., 0., 0.]
+        self.next_transform = None
+
+        # Register gcode commands
+        self.gcode.register_command('SET_Z_THERMAL_ADJUST',
+                                    self.cmd_SET_Z_THERMAL_ADJUST,
+                                    desc=self.cmd_SET_Z_THERMAL_ADJUST_help)
+        self.gcode.register_command('QUERY_Z_THERMAL_ADJUST',
+                                    self.cmd_QUERY_Z_THERMAL_ADJUST,
+                                    desc=self.cmd_QUERY_Z_THERMAL_ADJUST_help)
+        self.gcode.register_command('SET_Z_THERMAL_ADJUST_REF',
+                                    self.cmd_SET_Z_THERMAL_ADJUST_REF,
+                                    desc=self.cmd_SET_Z_THERMAL_ADJUST_REF_help)
+
+    def handle_connect(self):
+        'Called after all printer objects are instantiated'
+        self.toolhead = self.printer.lookup_object('toolhead')
+        gcode_move = self.printer.lookup_object('gcode_move')
+
+        # Temperature sensor config check and callback registration
+        try:
+            self.sensor = self.printer.lookup_object(self.temp_sensor_name)
+        except Exception as e:
+            msg = '''
+            %s\nUse the full config section name in the "temp_sensor" parameter,
+            e.g. "temp_sensor: temperature_sensor frame".
+            ''' % e
+            raise self.printer.config_error(msg)
+        else:
+            self.sensor.sensor.setup_callback(self.z_adj_temperature_callback)
```

Comment:
It's not valid to "peak" into another class like this (see docs/Code_Overview.md ).  I don't undersand why this is being done.  Shouldn't the code instantiate a new temperature sensor like other temperature modules?

Reply:
I legit forgot that I did this. It's a workaround; I originally implemented it as you described, but people complained they couldn't see it in their front end, use it with a temperature_fan, etc. I think `duplicate_pin_override` might be a viable workaround for those people now, no need to do it in here.